### ISSUE TITLE
Optimize categorical crossentropy gradient update

### ISSUE THREAD

**Issue by NicolasHug** [ogrisel/pygbm#91]
The gradient and hessian update of the categorical crossentropy loss  computes `p_k`  k times, but it only needs to compute it once (see https://github.com/scikit-learn/scikit-learn/commit/9e68984f8066085bdf801ccd0b5edca968709b5f which led to serious improvement in fit time).

**Comment by ogrisel** [comment]
+1 as well.

**Comment by aditya1702** [comment]
@NicolasHug Can I also work on this project?

**Comment by ogrisel** [comment]
@aditya1702 please feel free to submit a PR.
